scala/lambdas/ingest-event-aggregator/src/main/scala/uk/gov/nationalarchives/eventaggregator/Aggregator.scala

```diff
@@ -0,0 +1,104 @@
+
+  given aggregator[F[_]: Async: Generators: DASFNClient: DADynamoDBClient: Parallel]: Aggregator[F] = new Aggregator[F]:
+    private val logger: SelfAwareStructuredLogger[F] = Slf4jFactory.create[F].getLogger
+
+    private def logWithReason(sourceId: String)(newBatchReason: NewBatchReason): F[Unit] =
+      logger.info(Map("action" -> "Start new batch", "reason" -> newBatchReason.toString, "sourceId" -> "sourceId"))("Starting a new batch")
+
+    private def toDynamoString(value: String): AttributeValue = AttributeValue.builder.s(value).build
+
+    def writeToLockTable(input: Input, config: Config, batchId: String)(using dynamoClient: DADynamoDBClient[F]): F[Int] = {
+      dynamoClient.writeItem(
+        DADynamoDbWriteItemRequest(
+          config.lockTable,
+          Map("messageId" -> toDynamoString(input.id.toString), "batchId" -> toDynamoString(batchId), "message" -> toDynamoString(input.asJson.printWith(Printer.noSpaces)))
+        )
+      )
+    }
+
+    private def startNewBatch(ref: Ref[F, Map[String, Batch]], sourceId: String, config: Config, lambdaTimeoutTime: Long)(using
+        sfnClient: DASFNClient[F],
+        enc: Encoder[SFNArguments]
+    ): F[String] = {
+      val expires = lambdaTimeoutTime + (config.maxSecondaryBatchingWindow * 1000)
+      val delay = Math.ceil((expires - Generators[F].generateInstant.toEpochMilli).toDouble / 1000).toInt
+      val groupId = s"${config.sourceSystem}_${Generators[F].generateUuid}"
+      val batchId = s"${groupId}_0"
+      for {
+        batchCache <- ref.update(batchCache => batchCache + (sourceId -> Batch(batchId, Instant.ofEpochMilli(expires), 1)))
+        _ <- sfnClient.startExecution(config.sfnArn, SFNArguments(groupId, batchId, delay), batchId.some)
+      } yield batchId
+    }
+
+    private def getBatchId(ref: Ref[F, Map[String, Batch]], config: Config, sourceId: String, lambdaTimeoutTime: Long)(using Encoder[SFNArguments]): F[String] = {
+      def log = logWithReason(sourceId)
+      ref.get.flatMap { batchCache =>
+        val potentialCurrentBatchForSource = batchCache.get(sourceId)
+        potentialCurrentBatchForSource match
+          case None => log(NoExistingBatch) >> startNewBatch(ref, sourceId, config, lambdaTimeoutTime)
+          case Some(currentBatchForSource) =>
+            if lambdaTimeoutTime >= currentBatchForSource.expires.toEpochMilli then log(ExpiryBeforeLambdaTimeout) >> startNewBatch(ref, sourceId, config, lambdaTimeoutTime)
+            else if currentBatchForSource.items > config.maxBatchSize then log(MaxBatchSizeExceeded) >> startNewBatch(ref, sourceId, config, lambdaTimeoutTime)
+            else ref.update(_ + (sourceId -> currentBatchForSource.copy(items = currentBatchForSource.items + 1))) >> Async[F].pure(currentBatchForSource.batchId)
+      }
+    }
+
+    override def aggregate(config: Config, ref: Ref[F, Map[String, Batch]], messages: List[SQSMessage], remainingTimeInMillis: Int)(using
```

Comment:
Because we can't change the Lambda class declaration or the `override def handleRequest(input: SQSEvent, context: Context)` method it makes it tricky to test. We could put it in the Lambda class but then we'd have to test the Lambda class which is tricky so it feels easier to put all the logic in another class and the handleRequest method just calls that.